core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1180,26 +1202,36 @@ private int reuseOrCreateNewSortOrderId(SortOrder newOrder) {
       return newOrderId;
     }
 
-    private void setCurrentSnapshot(Snapshot snapshot, Long currentTimestampMillis) {
-      if (snapshot == null) {
-        this.currentSnapshotId = -1;
-        snapshotLog.clear();
-        changes.add(new MetadataUpdate.SetCurrentSnapshot(null));
-        return;
-      }
-
-      if (currentSnapshotId == snapshot.snapshotId()) {
-        return;
+    private void setBranchSnapshot(Snapshot snapshot, String branch, Long currentTimestampMillis) {
+      SnapshotRef ref = refs.get(branch);
+      if (ref != null) {
+        ValidationException.check(ref.isBranch(), "Cannot update branch: %s is a tag", branch);
+        if (ref.snapshotId() == snapshot.snapshotId()) {
+          return;
+        }
       }
 
       ValidationException.check(formatVersion == 1 || snapshot.sequenceNumber() <= lastSequenceNumber,
           "Last sequence number %s is less than existing snapshot sequence number %s",
           lastSequenceNumber, snapshot.sequenceNumber());
 
       this.lastUpdatedMillis = currentTimestampMillis != null ? currentTimestampMillis : snapshot.timestampMillis();
-      this.currentSnapshotId = snapshot.snapshotId();
-      snapshotLog.add(new SnapshotLogEntry(lastUpdatedMillis, snapshot.snapshotId()));
-      changes.add(new MetadataUpdate.SetCurrentSnapshot(snapshot.snapshotId()));
+
+      if (SnapshotRef.MAIN_BRANCH.equals(branch)) {
+        this.currentSnapshotId = snapshot.snapshotId();
+        snapshotLog.add(new SnapshotLogEntry(lastUpdatedMillis, snapshot.snapshotId()));
+      }
+
+      SnapshotRef.Builder refBuilder = SnapshotRef.branchBuilder(snapshot.snapshotId());
```

Comment:
I think you can just call `SnapshotRef.builderFrom` here

Reply:
The problem is that `ref` may be null. If it is null, then we create a new one with defaults. If it is non-null, then we copy it. I'll make this a bit more clear.